### ISSUE TITLE
Virtual Services ignored in Istio canary

### ISSUE THREAD

**Issue by Adharsh-Muraleedharan** [istio/istio#26104]
(NOTE: This is used to report product bugs:
  To report a security vulnerability, please visit <https://istio.io/about/security-vulnerabilities/>
  To ask questions about how to use Istio, please visit <https://discuss.istio.io>
)

**Bug description**
When doing a canary upgrade on ISTIO from version 1.5.0 to 1.6.6 we are facing issue with the virtualservices that are added is not getting reflected in the proxy routes. We can see the virtualservice routes no present in the istioctl proxy-config routes. All request reaching ingress gateway gets a 404

**Affected product area (please put an X in all that apply)**

[ ] Docs
[ ] Installation
[ X] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Expected behavior**
To upgarde to new control plane should automatically pick up the existing virtualservice routes as well as the new virtualservice routes that we apply

**Steps to reproduce the bug**
From version 1.5.0 upgrade to 1.6.6 using the below commands
**To install new canary deployment:**
curl -L https://git.io/getLatestIstio | ISTIO_VERSION=1.6.6 sh -
cd istio-1.6.6/bin
./istioctl install --set revision=canary

**To route to new canary deployment**
kubectl label namespace #{namespace_name} istio-injection- istio.io/rev=canary
kubectl rollout restart deployment -n #{namespace}

checking after applying a virtualservice it doesnt appear in the routes config

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
ISTIOCTL 1.6.6

**How was Istio installed?**
Using istioctl

**Environment where bug was observed (cloud vendor, OS, etc)**
AWS EKS

Additionally, please consider attaching a [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) by attaching
the dump file to this issue.

**Comment by esnible** [comment]
I labeled this `area/networking` because it seems to be a problem of Virtual Service rules being ignored, not a problem with Istio being installed.  If it belongs to another group please assign to another working group.

**Comment by VinothChinnadurai** [comment]
@esnible can you help you how to override the helm charts & values of istio 1.6 to support istio IG headless service, so that it wont go and override our existing setup

**Comment by esnible** [comment]
I don't know how to override the install to fix a problem with missing VirtualService routes.

Run `istioctl analyze` to see if anything is broken in the environment.
Run `istioctl proxy-status` and check if any pods are `STALE`.
If neither of those things is true, Istio 1.6.x should have identical behavior to 1.5.0.  Why do you think the upgrade changed something?  Do you suspect a root cause?

**Comment by VinothChinnadurai** [comment]
@ensible @gargnupur @ramaraochavali  Sorry for the delay. Here I am sharing the configuration

cd istio-1.5.0/
```
./istioctl manifest generate \
     --set values.gateways.istio-ingressgateway.type=ClusterIP \
     --set values.grafana.enabled=false \
     --set values.tracing.enabled=true \
     --set values.kiali.enabled=true \
     --set values.prometheus.enabled=false \
     --set values.global.proxy.accessLogFile="/dev/stdout" \
     --set values.kiali.dashboard.jaegerURL=http://jaeger-query:16686 > $HOME/install_istio.yaml
```
Then we opened `install_istio.yaml` and then edited  Istio IG service as `ClusterIP: None` and   `external-dns.alpha.kubernetes.io/hostname: xxx.xxx.xx`

Now we tried to upgrade in canary way.
cd istio-1.6.6/
```
./istioctl manifest generate \
     --set values.gateways.istio-ingressgateway.type=ClusterIP \
     --set values.grafana.enabled=false \
     --set values.tracing.enabled=true \
     --set values.kiali.enabled=true \
     --set values.prometheus.enabled=false \
     --set values.meshConfig.accessLogFile="/dev/stdout" \
     --set values.kiali.dashboard.jaegerURL=http://jaeger-query:16686 \
     --set revision=canary > $HOME/install_istio_1.6.6.yaml
```

Then we opened `install_istio_1.6.6.yaml` and then edited  Istio IG service as `ClusterIP: None` and   `external-dns.alpha.kubernetes.io/hostname: xxx.xxx.xx` then applied the same.

After applying 1.6.6 i can see the below list of pods in istio-system
```
istio-ingressgateway-5b95b9fdf7-jlwvp   1/1     Running   
istio-tracing-8584b4d7f9-5hsxh          1/1     Running   
istiod-5766465646-5d5d5                 1/1     Running   
istiod-canary-65dc749467-tw768          1/1     Running   
jaeger-6f457f5964-wgj8c                  1/1     Running   
```
I can see two pods .. one as istiod & istiod-canary but istio ingressgateway also get restarted which resulted in all routes got cleared.. We can see all virtual services present `kubectl get vs`

I tried to see routes..all routes were cleared which results 404.
```./istioctl proxy-config routes istio-ingressgateway-598796f4d9-xlkx4 -n istio-system -o json```

We are trying to understand the issue on why all routes are cleared and are there a way to mitigate from this?

**Comment by esnible** [comment]
Does `istioctl ps` show any pods being stale?

**Comment by VinothChinnadurai** [comment]
@esnible 
We tried the following
cd istio-1.6.6/
```
./istioctl analyze -n test

Error [IST0101] (VirtualService sleep-vs.test) Referenced gateway not found: "ingressgateway.istio-system.svc.cluster.local"
Error: Analyzers found issues when analyzing namespace: test.
See https://istio.io/docs/reference/config/analysis for more information about causes and resolutions.
```

```
./istioctl proxy-status
NAME                                                   CDS        LDS        EDS        RDS          PILOT                              VERSION
istio-egressgateway-89444576-s7njl.istio-system        SYNCED     SYNCED     SYNCED     NOT SENT     istiod-canary-5457864ccf-g6wxb     1.6.6
istio-ingressgateway-6df7b486d9-tmkbd.istio-system     SYNCED     SYNCED     SYNCED     SYNCED       istiod-canary-5457864ccf-g6wxb     1.6.6
```

Kindly let me know how to overcome this?

**Comment by esnible** [comment]
@VinothChinnadurai The error message may not be valid.  Use `--all-namespaces` and see if it goes away.

> ```
> ./istioctl analyze -n test
> 
> Error [IST0101] (VirtualService sleep-vs.test) Referenced gateway not found: "ingressgateway.istio-system.svc.cluster.local"
> Error: Analyzers found issues when analyzing namespace: test.
> ```

When _istioctl_ is run with `-n <namespace>` it may only check that namespace and assume `istio-system` is using only the defaults.

**Comment by VinothChinnadurai** [comment]
@esnible 
```
Error [IST0101] (VirtualService sleep-vs.test) Referenced gateway not found: "ingressgateway.istio-system.svc.cluster.local"
Warn [IST0102] (Namespace aws-sb) The namespace is not enabled for Istio injection. Run 'kubectl label namespace aws-sb istio-injection=enabled' to enable it, or 'kubectl label namespace aws-sb istio-injection=disabled' to explicitly mark it as not needing injection
Warn [IST0102] (Namespace catalog) The namespace is not enabled for Istio injection. Run 'kubectl label namespace catalog istio-injection=enabled' to enable it, or 'kubectl label namespace catalog istio-injection=disabled' to explicitly mark it as not needing injection
Warn [IST0102] (Namespace default) The namespace is not enabled for Istio injection. Run 'kubectl label namespace default istio-injection=enabled' to enable it, or 'kubectl label namespace default istio-injection=disabled' to explicitly mark it as not needing injection
Warn [IST0102] (Namespace external-secrets) The namespace is not enabled for Istio injection. Run 'kubectl label namespace external-secrets istio-injection=enabled' to enable it, or 'kubectl label namespace external-secrets istio-injection=disabled' to explicitly mark it as not needing injection
Warn [IST0102] (Namespace kube-node-lease) The namespace is not enabled for Istio injection. Run 'kubectl label namespace kube-node-lease istio-injection=enabled' to enable it, or 'kubectl label namespace kube-node-lease istio-injection=disabled' to explicitly mark it as not needing injection
Warn [IST0102] (Namespace kubecost) The namespace is not enabled for Istio injection. Run 'kubectl label namespace kubecost istio-injection=enabled' to enable it, or 'kubectl label namespace kubecost istio-injection=disabled' to explicitly mark it as not needing injection
Warn [IST0104] (Gateway kiali-gateway.istio-system) The gateway refers to a port that is not exposed on the workload (pod selector istio=ingressgateway; port 15029)
Warn [IST0104] (Gateway tracing-gateway.istio-system) The gateway refers to a port that is not exposed on the workload (pod selector istio=ingressgateway; port 15032)
Info [IST0118] (Service aws-servicebroker-aws-servicebroker.aws-sb) Port name  (port: 443, targetPort: 3199) doesn't follow the naming convention of Istio port.
Info [IST0118] (Service catalog-catalog-controller-manager.catalog) Port name secure (port: 443, targetPort: 8444) doesn't follow the naming convention of Istio port.
Info [IST0118] (Service catalog-catalog-webhook.catalog) Port name secure (port: 443, targetPort: 8443) doesn't follow the naming convention of Istio port.
Info [IST0118] (Service external-secrets-kubernetes-external-secrets.external-secrets) Port name prometheus (port: 3001, targetPort: prometheus) doesn't follow the naming convention of Istio port.
Info [IST0118] (Service kubecost-cost-analyzer.kubecost) Port name frontend (port: 9090, targetPort: 9090) doesn't follow the naming convention of Istio port.
Info [IST0118] (Service kubecost-cost-analyzer.kubecost) Port name model (port: 9003, targetPort: 9003) doesn't follow the naming convention of Istio port.
Info [IST0118] (Service kubecost-cost-analyzer.kubecost) Port name server (port: 9001, targetPort: 9001) doesn't follow the naming convention of Istio port.
Info [IST0118] (Service kubecost-grafana.kubecost) Port name service (port: 80, targetPort: 3000) doesn't follow the naming convention of Istio port.
Info [IST0118] (Service kubecost-prometheus-node-exporter.kubecost) Port name metrics (port: 9100, targetPort: 9100) doesn't follow the naming convention of Istio port.
Error: Analyzers found issues when analyzing all namespaces.
See https://istio.io/docs/reference/config/analysis for more information about causes and resolutions
```

**Comment by howardjohn** [comment]
I think this is fixed by #26352 in 1.6.8+

**Comment by howardjohn** [comment]
Please re-open if https://github.com/istio/istio/pull/26352 does not fix this. thanks!
